File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        result = []

        length = len(nums)
        for i in range(length - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            j = i + 1
            k = length - 1
            while j < k:
                current_sum = nums[i] + nums[j] + nums[k]
                if current_sum > 0:
                    k -= 1
                elif current_sum < 0:
                    j += 1
                else:
                    result.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                    while k > j and nums[k] == nums[k+1]:
                        k -= 1    
        return result                             
        

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-3.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> prefixSums = new HashMap<>();
        prefixSums.put(0, 1);
        int currentPrefix = 0, result = 0;

        for(int num : nums) {
            currentPrefix += num;
            int needed = currentPrefix - k;
            result += prefixSums.getOrDefault(needed, 0);
            prefixSums.put(currentPrefix, 1 + prefixSums.getOrDefault(currentPrefix, 0));
        }

        return result;
    }
}